staker/fast_confirm.go

```diff
@@ -0,0 +1,249 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/offchainlabs/nitro/solgen/go/contractsgen"
+	"github.com/offchainlabs/nitro/solgen/go/rollupgen"
+	"github.com/offchainlabs/nitro/staker/txbuilder"
+	"github.com/offchainlabs/nitro/util/headerreader"
+)
+
+type FastConfirmSafe struct {
+	rollupAddress             common.Address
+	safe                      *contractsgen.Safe
+	owners                    []common.Address
+	threshold                 uint64
+	fastConfirmNextNodeMethod abi.Method
+	builder                   *txbuilder.Builder
+	wallet                    ValidatorWalletInterface
+	gasRefunder               common.Address
+	l1Reader                  *headerreader.HeaderReader
+	fastConfirmApprover       common.Address
+}
+
+func NewFastConfirmSafe(
+	callOpts bind.CallOpts,
+	rollupAddress common.Address,
+	fastConfirmSafeAddress common.Address,
+	builder *txbuilder.Builder,
+	wallet ValidatorWalletInterface,
+	gasRefunder common.Address,
+	l1Reader *headerreader.HeaderReader,
+	fastConfirmApprover common.Address,
+) (*FastConfirmSafe, error) {
+	fastConfirmSafe := &FastConfirmSafe{
+		builder:             builder,
+		rollupAddress:       rollupAddress,
+		wallet:              wallet,
+		gasRefunder:         gasRefunder,
+		l1Reader:            l1Reader,
+		fastConfirmApprover: fastConfirmApprover,
+	}
+	safe, err := contractsgen.NewSafe(fastConfirmSafeAddress, builder)
+	if err != nil {
+		return nil, err
+	}
+	fastConfirmSafe.safe = safe
+	owners, err := safe.GetOwners(&callOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	// This is needed because safe contract needs owners to be sorted.
+	sort.Slice(owners, func(i, j int) bool {
+		return owners[i].Cmp(owners[j]) < 0
+	})
+	fastConfirmSafe.owners = owners
+	threshold, err := safe.GetThreshold(&callOpts)
+	if err != nil {
+		return nil, err
+	}
+	fastConfirmSafe.threshold = threshold.Uint64()
+	rollupUserLogicAbi, err := rollupgen.RollupUserLogicMetaData.GetAbi()
+	if err != nil {
+		return nil, err
+	}
+	fastConfirmNextNodeMethod, ok := rollupUserLogicAbi.Methods["fastConfirmNextNode"]
+	if !ok {
+		return nil, errors.New("RollupUserLogic ABI missing fastConfirmNextNode method")
+	}
+	fastConfirmSafe.fastConfirmNextNodeMethod = fastConfirmNextNodeMethod
+	return fastConfirmSafe, nil
+}
+
+func (s *Staker) tryFastConfirmationNodeNumber(ctx context.Context, number uint64) error {
+	if !s.config.EnableFastConfirmation {
+		return nil
+	}
+	nodeInfo, err := s.rollup.LookupNode(ctx, number)
+	if err != nil {
+		return err
+	}
+	return s.tryFastConfirmation(ctx, nodeInfo.AfterState().GlobalState.BlockHash, nodeInfo.AfterState().GlobalState.SendRoot)
+}
+
+func (s *Staker) tryFastConfirmation(ctx context.Context, blockHash common.Hash, sendRoot common.Hash) error {
```

Comment:
Let's keep the Staker methods in `staker.go` and only have the FastConfirmSafe methods in this file

Reply:
Done.